core/broadleaf-framework/src/main/java/org/broadleafcommerce/core/search/service/solr/indexer/AbstractSolrIndexUpdateServiceImpl.java

```diff
@@ -0,0 +1,292 @@
+            } else {
+                LOG.warn("A command thread has already been registered for the following command group: " + getCommandGroup());
+            }
+        }
+    }
+    
+    @Override
+    public void destroy() throws Exception {
+        synchronized (AbstractSolrIndexUpdateServiceImpl.class) {
+            AtomicReferenceArray<Runnable> runnables = commandThreadRegistry.remove(getCommandGroup());
+            if (runnables != null) {
+                ((CommandCoordinator)runnables.get(0)).stopRunning();
+                ((Thread)runnables.get(1)).interrupt();
+            }
+        }
+    }
+    
+    /**
+     * Stops all threads that are listening to various command queues.
+     */
+    public static void shutdownAll() {
+        synchronized (AbstractSolrIndexUpdateServiceImpl.class) {
+            Set<Entry<String, AtomicReferenceArray<Runnable>>> entries = commandThreadRegistry.entrySet();
+            for (Entry<String, AtomicReferenceArray<Runnable>> entry : entries) {
+                ((CommandCoordinator)entry.getValue().get(0)).stopRunning();
+                ((Thread)entry.getValue().get(1)).interrupt();
+            }
+            
+            commandThreadRegistry.clear();
+        }
+    }
+    
+    /**
+     * This is any arbitrary name to identify or group commands, typically based on the Solr index (or indexes) being updated.  
+     * For example, "catalog" will likely be a command identifier.  It could also be "product", but in less common cases where you want to 
+     * index categories or other things with products and you want to serialize those commands, then this can assist.
+     * 
+     * @return
+     */
+    public final String getCommandGroup() {
+        return commandGroup;
+    }
+    
+    protected final <C extends SolrUpdateCommand> void scheduleCommand(C command) {
+        if (!isRunning(getCommandGroup())) {
+            throw new IllegalStateException("Attempted to queue a SolrUpdateCommand but " + getClass().getName() + " is not initialized or has been shut down.");
+        }
+        if (command != null) {
+            try {
+                for (int i = 0; i < 5; i++) {
+                    // Assuming there is not already an equal command in the queue, and not a FullReindexCommand in the queue, then add the command
+                    // for processing in a background thread.
+                    // If a FullReindexCommand is in the queue, then all incremental update commands are ignored since they'll
+                    // be overridden when the full reindex is applied. Just before the FullReindexCommand begins, it
+                    // is removed from the queue.  This allows incremental update commands to be queued with the assumption
+                    // that the full reindex may be mid-stream & may not have picked up the update.
+                    if (!commandQueue.contains(FullReindexCommand.DEFAULT_INSTANCE) && !commandQueue.contains(command)) {
```

Comment:
@ktisdell I'm beginning to question this condition - do we really want to avoid queueing incremental update commands if a FullReindexCommand is in queue?

I think that logic is totally fine if the full reindex is successful, but if the full reindex fails, then you're left with a foreground collection that won't be updated.

On the other hand, if we **do** queue the incremental updates anyway, then they'll repeat a bit of the same work, but it _should be_ harmless.

Reply:
@ckittrell I don't disagree and don't have a strong opinion here.  I think more work could be done to have a coordinated foreground command queue and background command queue which could effectively handle situations like that.  Given that we don't, I'm OK with removing that logic in favor of "the extra work" that will have to be done by the indexer, assuming that doing that extra work simply restores the state after it is complete (i.e. causes no harm).

Reply:
Agreed - lets hit the easy button here & remove the `!commandQueue.contains(FullReindexCommand.DEFAULT_INSTANCE)` portion of the condition.